filament/src/ShadowMapManager.cpp

```diff
@@ -91,31 +91,71 @@ void ShadowMapManager::prepare(FEngine& engine, DriverApi& driver, SamplerGroup&
 
     // If we already have a texture with the same dimensions and layer count, there's no need to
     // create a new one.
-    const TextureState newState(dim, layersNeeded);
+    const TextureState newState(dim, layersNeeded, mUseVsm);
     if (mTextureState == newState) {
         // nothing to do here.
-        assert(mShadowMapTexture);
+        assert(mShadowMapTexture || mVsmTexture);
         return;
     }
 
     // destroy the current rendertargets and texture
     destroyResources(driver);
 
-    mShadowMapTexture = driver.createTexture(
-            SamplerType::SAMPLER_2D_ARRAY, 1, mTextureFormat, 1, dim, dim, layersNeeded,
-            TextureUsage::DEPTH_ATTACHMENT |TextureUsage::SAMPLEABLE);
+    // Shadows are either all PCF or all VSM.
+
+    if (mUseVsm) {
+        // TODO: support 16-bit VSM depth textures.
```

Comment:
If I remember correctly we can get away with using just the 16 bit variant right?

Reply:
I haven't tried that yet, but it's on my list. Yeah, if 16 bits is sufficient, then we could just use that unconditionally.

Reply:
https://mynameismjp.wordpress.com/2013/09/10/shadow-maps/ has info on that I believe